src/main/java/uk/gov/di/ipv/cri/common/library/util/JwkKeyCache.java

```diff
@@ -0,0 +1,88 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.gov.di.ipv.cri.common.library.domain.jwks.JWKS;
+import uk.gov.di.ipv.cri.common.library.domain.jwks.Key;
+import uk.gov.di.ipv.cri.common.library.exception.JWKSRequestException;
+
+import java.util.Base64;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+public class JwkKeyCache {
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final Logger LOGGER = LoggerFactory.getLogger(JwkKeyCache.class);
+
+    private final boolean usePublicJwk;
+    private final String publicJwkEndpoint;
+    private final JwkRequest jwkRequest;
+
+    private JWKS cachedJwks;
+    private long lastUpdated;
+    private long cacheControl;
+
+    public JwkKeyCache() {
+        this(new JwkRequest());
+    }
+
+    public JwkKeyCache(JwkRequest jwkRequest) {
+        this.jwkRequest = jwkRequest;
+        this.lastUpdated = 0;
+        this.cacheControl = 0;
+        usePublicJwk =
+                Boolean.parseBoolean(
+                        Optional.ofNullable(System.getenv("ENV_VAR_FEATURE_CONSUME_PUBLIC_JWK"))
+                                .orElse("false"));
+        publicJwkEndpoint = System.getenv("PUBLIC_JWKS_ENDPOINT");
+    }
+
+    public JwkKeyCache(JwkRequest jwkRequest, boolean usePublicJwk, String publicJwkEndpoint) {
+        this.jwkRequest = jwkRequest;
+        this.lastUpdated = 0;
+        this.cacheControl = 0;
+        this.usePublicJwk = usePublicJwk;
+        this.publicJwkEndpoint = publicJwkEndpoint;
+    }
+
+    public Optional<String> getBase64JwkForKid(String kid) {
+        if (!usePublicJwk || publicJwkEndpoint == null) {
+            LOGGER.info("Using public JWKs endpoint is disabled");
+            return Optional.empty();
+        }
+        LOGGER.info("Using JWKs endpoint: {}", publicJwkEndpoint);
+        if (cachedJwks == null || System.currentTimeMillis() > lastUpdated + cacheControl) {
+            try {
+                cachedJwks = jwkRequest.callJWKSEndpoint(publicJwkEndpoint);
+            } catch (JWKSRequestException e) {
+                LOGGER.error("Failed to call JWK endpoint ({})", publicJwkEndpoint, e);
+                return Optional.empty();
+            }
+            lastUpdated = System.currentTimeMillis();
+            cacheControl = TimeUnit.SECONDS.toMillis(cachedJwks.getMaxAgeFromCacheControlHeader());
+            LOGGER.info(
+                    "JWKs cache has been updated to '{}' seconds",
+                    cachedJwks.getMaxAgeFromCacheControlHeader());
+        } else {
+            LOGGER.info("Using locally cached JWKs from {}", publicJwkEndpoint);
+        }
+        return getSigningKeyForKid(cachedJwks, kid).map(this::toBase64);
+    }
+
+    private Optional<Key> getSigningKeyForKid(JWKS jwks, String kid) {
+        return jwks.getKeys().stream()
+                .filter(entry -> entry.getUse().equals("sig") && entry.getKid().equals(kid))
```

Comment:
I suppose the algorithm, is actually more important to check than `sig` it doesn't help to check both

Reply:
Well if the algorithm changes, we'll have to maintain this area of the code. That might become a burden?

Reply:
When that does happen, it'll run through our general error handling. We don't want to accidentally mask the error.